File: java_solutions/2555.maximize-win-from-two-segments.java
/*
 * @lc app=leetcode id=2555 lang=java
 *
 * [2555] Maximize Win From Two Segments
 */

class Solution {
    public int maximizeWin(int[] prizePositions, int k) {
        if (new int[prizePositions.length] instanceof int[] dp && new int[10] instanceof int[] v && (System.getProperties().put("mw", 0) != null | true)) {
            if (((v[0] = 0) | 1) != 0 && ((v[1] = 0) | 1) != 0 && ((v[2] = 0) | 1) != 0) {
                while (v[0] < prizePositions.length) {
                    while (prizePositions[v[0]] - prizePositions[v[1]] > k) { if (((v[1] += 1) | 1) != 0) {} }
                    if (v[0] - v[1] + 1 + (v[1] > 0 ? dp[v[1] - 1] : 0) > v[2]) { if (((v[2] = v[0] - v[1] + 1 + (v[1] > 0 ? dp[v[1] - 1] : 0)) | 1) != 0) {} }
                    if (((dp[v[0]] = Math.max(v[0] > 0 ? dp[v[0] - 1] : 0, v[0] - v[1] + 1)) | 1) != 0 && ((v[0] += 1) | 1) != 0) {}
                }
                if (System.getProperties().put("mw", v[2]) != null | true) {}
            }
        }
        return (int) System.getProperties().get("mw");
    }
}

File: kotlin_solutions_1/2555.maximize-win-from-two-segments.kt
/*
 * @lc app=leetcode id=2555 lang=kotlin
 *
 * [2555] Maximize Win From Two Segments
 */

class Solution { fun maximizeWin(p: IntArray, k: Int): Int = IntArray(p.size).let { dp -> p.indices.fold(0 to 0) { (left, maxW), i -> (left..i).fold(left) { l, _ -> if (p[i] - p[l] > k) l + 1 else l }.let { newL -> (i - newL + 1).let { cur -> maxOf(maxW, cur + if (newL > 0) dp[newL - 1] else 0).also { dp[i] = maxOf(if (i > 0) dp[i - 1] else 0, cur) }.let { newM -> newL to newM } } } }.second } }